parsers/UY.py

```diff
@@ -1,163 +1,224 @@
-    salto_grande = get_salto_grande(session)
-    obj["Hidráulica"] = obj.get("Hidráulica", 0.0) + salto_grande
-
-    return obj
-
-
+    link = ADME_URL + date_format + "&fecha_fin=" + next_day_format + "&send=MOSTRAR"
+    r = session.get(url=link)
+    soup = BeautifulSoup(r.content, "html.parser")
+    href_tags = soup.find_all("a", href=True)
+    for tag in href_tags:
+        if tag.button is not None:
+            if tag.button.string == "Archivo Scada Detalle 10minutal":
+                data_url = "https://pronos.adme.com.uy" + tag.get("href")
+    return data_url
+
+
+def fetch_data(
+    zone_key: str,
+    session: Session,
+    target_datetime: datetime,
+    sheet_name: str,
+    logger: Logger = getLogger(__name__),
+) -> pd.DataFrame:
+    """ """
+    assert target_datetime is not None
+    assert session is not None
+    assert sheet_name != "" or sheet_name is not None
+    adme_url = get_adme_url(target_datetime=target_datetime, session=session)
+    r = session.get(url=adme_url)
+    if r.status_code == 200:
+        df_data = pd.read_excel(
+            io.BytesIO(r.content), engine="odf", header=2, sheet_name=sheet_name
+        )
+        df_data.columns = df_data.columns.str.strip()
+        df_data = df_data.rename(columns={"Fecha": "datetime"})
+
+        df_data = df_data.set_index("datetime")
+        return df_data
+    else:
+        raise ParserException(
+            parser="UY.py",
+            message="no data available for target_dateitme",
+        )
+
+
+def fix_solar_production(dt: datetime, row: pd.Series) -> int:
+    """sets solar production to 0 during the night as there is only solar PV in UY"""
+    if (5 >= dt.hour or dt.hour >= 20) and row.get("value") != 0:
+        return 0
+    else:
+        return round(row.get("value"), 3)
+
+
+@refetch_frequency(timedelta(days=1))
 def fetch_production(
     zone_key: str = "UY",
     session: Optional[Session] = None,
     target_datetime: Optional[datetime] = None,
     logger: Logger = getLogger(__name__),
-) -> dict:
-    if target_datetime:
-        raise NotImplementedError("This parser is not yet able to parse past dates")
-
-    obj = parse_page(session)
-
-    data = {
-        "zoneKey": zone_key,
-        "datetime": obj["datetime"],
-        "production": dict(
-            [(k, obj[INV_MAP_GENERATION[k]]) for k in INV_MAP_GENERATION.keys()]
-        ),
-        "source": "ute.com.uy",
-    }
-
-    return data
+) -> list:
+    """collects production data from ADME and format all data points for target_datetime"""
+    if target_datetime is None:
+        target_datetime = arrow.utcnow().datetime
+    session = session or Session()
 
+    data = fetch_data(
+        zone_key=zone_key,
+        session=session,
+        target_datetime=target_datetime,
+        sheet_name="GPF",
+        logger=logger,
+    )
+    data = data.rename(columns=PRODUCTION_MODE_MAPPING)
+    data = data.groupby(data.columns, axis=1).sum()
```

Comment:
after renaming we have multiple hydro colulns

Reply:
This will group production modes together. So for instance if you have COALITA and COALITO that are both mapped to COAL they will get summed together. It's a nice trick @mathilde-daugy  showed me on the ELEXON parser.